### PR TITLE
[Feature] Support native Apache Arrow format for Stream Load

### LINKED ISSUE #72069: [Feature] Support Apache Arrow format in Stream Load
- [x] https://github.com/StarRocks/starrocks/pull/74593

### Description

I would like to submit a feature request for **Stream Load** to support **Apache Arrow** as an input format.

Currently, when upstream systems already produce data in Arrow columnar format, the data still needs to be converted into CSV, JSON, or other intermediate formats before being loaded into StarRocks. This adds extra serialization, parsing, and memory copy overhead.

### Motivation

Supporting Arrow format in Stream Load could be useful for the following reasons:

- reduce unnecessary format conversion
- lower serialization and deserialization overhead
- improve ingestion efficiency in some scenarios
- make integration easier for systems that already use Arrow internally

### Possible Interface

For example, this could be exposed through a new format option:

```http
format: arrow
```

or by recognizing a content type such as:

```http
Content-Type: application/vnd.apache.arrow.stream
```

### Suggested Initial Scope

A first version may only need to support:

- Arrow Stream format
- Arrow RecordBatch input
- common primitive types

### Notes

- this should remain backward compatible with existing Stream Load formats
- schema mapping between Arrow schema and StarRocks table schema would be needed
- unsupported types or incompatible schemas should return clear error messages

### Typical Use Cases

This feature may be useful for:

- data integration tools
- ETL / ELT engines
- high-performance ingestion pipelines
- connectors that already use Arrow internally

### Non-goals for an Initial Version

- full Arrow Flight write protocol support
- all Arrow extension types
- changing existing Stream Load behavior

Thanks for considering this feature request.

### PR DESCRIPTION
**Pull request by jimmystewpot**
## Why I'm doing:

Currently, StarRocks lacks native Apache Arrow IPC Stream format support for Stream Load. Adding this capability allows users to stream Arrow data directly into StarRocks without converting to JSON/CSV, significantly improving ingestion throughput and performance.

| Metric | `BM_JsonScanner` | `BM_ArrowScanner` | Improvement / Impact |
| :--- | :--- | :--- | :--- |
| **Execution Time** | 674 ms | 49.2 ms | **13.7x Faster** (92.7% reduction in latency) |
| **Throughput** | 148.44k items/s | 2.03M items/s | **13.7x Throughput Increase** |
| **CPU Efficiency** | 674 ms | 49.2 ms | CPU time matches wall time; processing is heavily bound by computational/parsing efficiency rather than I/O wait. |

Depending on how this PR is received, I would also like to contribute arrow-ipc support for routine load as a follow-up PR.

### Note

This change is heavily based on the existing parquet feature. While reviewing the parquet feature, I encountered a possible bug. In https://github.com/StarRocks/starrocks/blob/c1d656a535228f7a6a0cedb60830d0bfc7b6dd5e/be/src/exec/file_scanner/parquet_scanner.cpp#L178, when a column is missing from the Arrow/Parquet file, the code `appends _batch->num_rows()` nulls instead of `num_elements` nulls. `num_elements` is the bounded count computed on lines 187-188 that respects `_max_chunk_size`. Using `_batch->num_rows()`. I wanted to flag this for review. I have included tests in this request for Arrow; however, I have not touched Parquet. I wanted to kick off a discussion, as I am happy to fix this if it is as I expect. Although I could be wrong.

## What I'm doing:

1. Implemented Native Arrow IPC Stream Format Parser/Scanner:
   - Registered `FORMAT_ARROW` in the HTTP stream load action, FileScanner, and connector dispatching.
   - Built `ArrowScanner` (inheriting from `FileScanner`) using the Arrow C++ library (`RecordBatchStreamReader`).

2. Hardened Scanner Implementation & Addressed Code Review Items:
   - Corrected Null-Padding: Solved a bug where padding missing columns at chunk boundaries used the full batch row count instead of the sliced chunk count (`num_elements`).
   - Observability: Added `read_batch_ns` RawTimer and `VLOG_ROW` logs for chunk profiling; set up the `report_error_message` callback to pipe data quality errors to the load error file and rejected record sync daemon.
   - Progress Reporting: Added bytes-scanned progress reporting via `update_num_bytes_scan_from_source` in the Arrow custom input stream.
   - Safety & Forward Compatibility: Added empty-check guard for `broker_addresses`; replaced deprecated `ValueOrDie()` with `MoveValueUnsafe()`; corrected `closed()` status contract.
   - Maintainability: Encapsulated `StarRocksArrowInputStream` inside an anonymous namespace; cleaned up dead headers and reordered include directives to comply with BE coding standards.

3. Dependencies & Build Updates:
   - Arrow C++ Library: Upgraded from `19.0.1` to `24.0.0` to support modern IPC Stream functionalities. (v24 introduces bit-unpacking overhauls that we have seen 40-50% performance improvements on, v21 improved on simd vectorisation for byte-stream split encoding for floating point types and fixed-width reading speedups).
   - Java Arrow SDK: Aligned Java-side dependencies by upgrading versions to `18.0.0` in `format-sdk/pom.xml` and `java-extensions/pom.xml`. 

4. Comprehensive Unit Testing (`arrow_scanner_test.cpp`):
   - Added `TestScanArrowStream` for basic stream validation.
   - Added `TestScanArrowStreamMismatchAndCast` for schema casting.
   - Added `TestScanArrowStreamNullColumnChunkBoundary` for chunk boundary null padding checks.
   - Added `TestScanArrowStreamStrictModeQualityError` to verify strict mode validation and rejected record logging.
   - Added edge case tests: `TestScanArrowStreamEmptyFile` (0-rows), `TestScanArrowStreamNullable` (explicit nulls), and `TestScanArrowStreamMultiBatch` (multiple batches).

Fixes: #72069

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information (adds Support for Arrow Stream format type in stream load)
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [x] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [x] I have added documentation for my new feature or new function
  - [ ] This pr needs auto generate documentation
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 4.1
  - [ ] 4.0
  - [ ] 3.5

### CODE REVIEW

**Comment by chl-wxp** [comment]
good job

**Review by stdpain — DISMISSED** [review]

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by stdpain** [comment]
@jimmystewpot u could create a new PR to resolve this 
https://github.com/StarRocks/starrocks/blob/c1d656a535228f7a6a0cedb60830d0bfc7b6dd5e/be/src/exec/file_scanner/parquet_scanner.cpp#L178

File: thirdparty/vars.sh [hunk 1]
```diff
@@ -247,10 +247,10 @@ FLATBUFFERS_SOURCE=flatbuffers-1.10.0
...
 
 # arrow
-ARROW_DOWNLOAD="https://github.com/apache/arrow/archive/refs/tags/apache-arrow-19.0.1.tar.gz"
-ARROW_NAME="arrow-apache-arrow-19.0.1.tar.gz"
-ARROW_SOURCE="arrow-apache-arrow-19.0.1"
-ARROW_MD5SUM="8c5091da0f8fb41a47d7f4dad7b712df"
+ARROW_DOWNLOAD="https://github.com/apache/arrow/archive/refs/tags/apache-arrow-24.0.0.tar.gz"
+ARROW_NAME="arrow-apache-arrow-24.0.0.tar.gz"
+ARROW_SOURCE="arrow-apache-arrow-24.0.0"
+ARROW_MD5SUM="66c53bd00baa79034bd2ca167beea436"
```
Comment (kevincai) [inline thirdparty/vars.sh:253]: please open a separate PR for this thirdparty change, the PR should be as simple as possible with version bump and possibly compiling fixes, this will offload the heavy thirdparty rebuild ci workflow.

Comment (stdpain) [reply thirdparty/vars.sh:253]: arrow upgrade in https://github.com/StarRocks/starrocks/pull/75868

Comment (jimmystewpot) [reply thirdparty/vars.sh:253]: I'll sync the branch with main and remove the arrow dep upgrade from this pr entirely.

**Comment by jimmystewpot** [comment]
I have fixed up the issues from the codex comments, and removed the Apache arrow dep upgrade. The PR is ready for review again!

**Comment by jimmystewpot** [comment]
@stdpain I have created a separate PR for the parquet null bug that I raised. It can be found here: https://github.com/StarRocks/starrocks/pull/75981

**Comment by CelerData-Reviewer** [comment]
@codex review

**Review by stdpain — DISMISSED** [review]

File: be/src/connector/file_connector.cpp
Comment (kevincai) [inline be/src/connector/file_connector.cpp:1]: this file changed from mode 644 to mode 755. it shouldn't

File: fe/pom.xml
Comment (kevincai) [inline fe/pom.xml:1]: revert this mode change as well.

File: fe/fe-core/src/test/java/com/starrocks/load/LoadTest.java
Comment (kevincai) [inline fe/fe-core/src/test/java/com/starrocks/load/LoadTest.java:1]: ditto

**Review by kevincai — CHANGES_REQUESTED** [review]
please remove all the file mode change from 644 -> 755.

**Comment by jimmystewpot** [comment]
@kevincai the permissions changes have been reverted as requested.

**Review by kevincai — DISMISSED** [review]

**Review by kevincai — APPROVED** [review]

**Comment by CelerData-Reviewer** [comment]
@codex review

**Review by stdpain — APPROVED** [review]

**Comment by stdpain** [comment]
Very useful feature. Thanks for your contribution!
